### PR TITLE
Gracefully deal with dangling symlinks

### PR DESCRIPTION
**Pull request by jeremy**
Gracefully deal with dangling symlinks
* Don't add files that are dangling symlinks
* When watching a directory, ignore dangling symlinks rather than treating them as forever-stale

Add a watcher logging facility so such things can be diagnosed and debugged.

### CODE REVIEW

File: lib/spring/application.rb [hunk 1]
```diff
@@ -67,6 +67,11 @@ def initialized?
     def start_watcher
       @watcher = Spring.watcher
       @watcher.on_stale { state! :watcher_stale }
+
+      @watcher.on_debug do |message|
+        spring_env.log "[watcher:#{app_env}] #{message}"
+      end if @watcher.respond_to?(:on_debug)
```
Comment (jonleighton) [inline lib/spring/application.rb]: Why is the conditional needed?

File: lib/spring/watcher/abstract.rb [hunk 2]
```diff
@@ -36,14 +48,30 @@ def add(*items)
           end
         end
 
-        items = items.select(&:exist?)
+        items = items.select do |item|
+          if item.symlink?
+            item.readlink.exist?.tap do |exists|
+              debug do
+                "add: ignoring dangling symlink: #{item.inspect} -> #{item.readlink.inspect}"
+              end if !exists
```
Comment (jonleighton) [inline lib/spring/watcher/abstract.rb]: Not really a fan of this style of postfix conditional on a multi-line block personally (I think it's harder to spot what the conditional is doing...)

File: lib/spring/application.rb (hunk 1 above)
Comment (jeremy) [inline lib/spring/application.rb]: Only the poller watcher implements it. Calling it on the listen watcher would bomb.

Comment (jonleighton) [inline lib/spring/application.rb]: But it's defined in the abstract class which the listen watcher inherits from?

**Pull request merged**